added PHP and UTF-8 section

ylynfatt opened the pull request:
Added PHP and UTF-8 section from PHP Best Practices website originally written by Alex Cabal.

Closes #301

ylynfatt committed:
added PHP and UTF-8 section

ylynfatt committed:
declared utf charset in the connection string

Also updated code example to match phpbestpractices.org example

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+
+### There's no one-liner. Be careful, detailed, and consistent.
+
+UTF-8 in PHP sucks. Sorry.
```
Probably a little aggressive.

ylynfatt replied:
Shall we just remove that line then?

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+
+### UTF-8 at the PHP level
+
+The basic [string operations](http://php.net/manual/en/language.operators.string.php), like concatenating two strings and assigning strings to variables, don't need anything special for UTF-8. However most [string functions](http://php.net/manual/en/ref.strings.php), like [`strpos()`](http://php.net/manual/en/function.strpos.php) and [`strlen()`](http://php.net/manual/en/function.strlen.php), do  ...
```
Can you move links from inline to below?

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+
+The basic [string operations](http://php.net/manual/en/language.operators.string.php), like concatenating two strings and assigning strings to variables, don't need anything special for UTF-8. However most [string functions](http://php.net/manual/en/ref.strings.php), like [`strpos()`](http://php.net/manual/en/function.strpos.php) and [`strlen()`](http://php.net/manual/en/function.strlen.php), do  ...
+
+You must use the `mb_*` functions whenever you operate on a Unicode string. For example, if you use [`substr()`](http://php.net/manual/en/function.substr.php) on a UTF-8 string, there's a good chance the result will include some garbled half-characters. The correct function to use would be the multibyte counterpart, [`mb_substr()`](http://php.net/manual/en/function.mb-substr.php).
```
+1 for garbled.

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+
+Finally, many PHP functions that operate on strings have an optional parameter letting you specify the character encoding. You should always explicitly indicate UTF-8 when given the option. For example, [`htmlentities()`](http://php.net/manual/en/function.htmlentities.php) has an option for character encoding, and you should always specify UTF-8 if dealing with such strings.
+
+### UTF-8 at the MySQL level
```
At the "Database" level instead? Link to PostgreSQL for the sake of looking like we're shoving one specific DB system down peoples throats.

ylynfatt replied:
@philsturgeon where exactly did you want me to put the link to PostgreSQL? In the Further Reading section?

philsturgeon replied:
I hadnt noticed this was copied from the other site. What I was suggesting was that instead of saying "UTF-8 at the MySQL level" it should be "UTF-8 at the Database level" instead, then mention the link to how its done in MySQL and try to cover the same for Postgres the best you can too. 

phpbestpractices.org might want to word this like MySQL is the only database people will be using, but I would like the text to be more generic. 

If you dunno what im rambling on about I can just merge this and word it myself. Let me know. :)

ylynfatt replied:
@philsturgeon okay I understand your point. After doing some reading, I realize that `utf8mb4` is MySQL specific. Also I could be wrong but from my brief reading it doesn't seem as if the `charset` portion of the DSN in the MySQL example will work for PostgreSQL. Since you have an idea of how you'd like to word it, perhaps you can go ahead and add that bit about PostgreSQL. Thanks. :)

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+
+If your PHP script accesses MySQL, there's a chance your strings could be stored as non-UTF-8 strings in the database even if you follow all of the precautions above.
+
+To make sure your strings go from PHP to MySQL as UTF-8, make sure your database and tables are all set to the utf8mb4 character set and collation, and that you use the utf8mb4 character set in the PDO connection string. For an example, see the section on [connecting to and querying a MySQL database](https://phpbestpractices.org/#mysql). This is _critically important_.
```
`utf8mb4` is code in some places and text in others. Can you make it consistent?

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,105 @@
+// Connect to a database to store the transformed string
+// See the PDO example in this document for more information
+// Note the `set names utf8mb4` commmand!
+$link = new \PDO(   'mysql:host=your-hostname;dbname=your-db;charset=utf8mb4',
```
Can you bump that first argument down a line so the example is PSR-2?

philsturgeon commented:
This looks great but it has a little work to go on formatting and wording. Good job. :)

ylynfatt committed:
bump first argument down a line so the example is PSR-2

ylynfatt committed:
made utf8mb4 consistently represented as code

ylynfatt committed:
removed sentence as wording seemed a little too aggressive

ylynfatt committed:
moved inline links to Further Reading section

ylynfatt committed:
changed wording so it speaks to code example in this section

philsturgeon commented on _posts/05-05-01-PHP-and-UTF8.md:
```diff
@@ -0,0 +1,115 @@
+
+Additionally, you should use the `mb_internal_encoding()` function at the top of every PHP script you write (or at the top of your global include script), and the `mb_http_output()` function right after it if your script is outputting to a browser. Explicitly defining the encoding of your strings in every script will save you a lot of headaches down the road.
+
+Finally, many PHP functions that operate on strings have an optional parameter letting you specify the character encoding. You should always explicitly indicate UTF-8 when given the option. For example, `htmlentities()` has an option for character encoding, and you should always specify UTF-8 if dealing with such strings.
```
Should probably add a note here explaining that PHP 5.4 has made the encoding default to `'UTF-8'`, and is something you should look out for.

http://php.net/htmlentities

ylynfatt replied:
This has now been added.

philsturgeon commented:
Looking good. If you can sort out the word wrapping (120) then I'll merge this one.

ylynfatt committed:
fixed word wrapping (now 120 chars)

Also added note about default encoding for htmlentities and
htmlspecialchars and added link to php docs for htmlspecialchars

ylynfatt commented:
@philsturgeon Word wrapping change now done.
